File: PYTHON/CinemaBookingSystem.py
from tabulate import tabulate
from openpyxl import Workbook, load_workbook

import random

# Defining movies and their prices
movies = {}
movies["1"] = {"title": "Ant-Man And The Wasp: Quantumania", "Genre": "Action/Adventure", "price": 680}
movies["2"] = {"title": "Avatar: The Way Of Water", "Genre": "Action/Adventure", "price": 680}
movies["3"] = {"title": "Sound Of Silence", "Genre": "Action/Adventure", "price": 540}
movies["4"] = {"title": "The First Slam Dunk", "Genre": "Anime", "price": 480}
movies["5"] = {"title": "Oras De Peligro", "Genre": "Drama", "price": 380}

# Defining the seats avalability 
seats = {
    movie_num: {
        "A": [i for i in range(1, 6)],
        "B": [i for i in range(1, 6)],
        "C": [i for i in range(1, 6)],
        "D": [i for i in range(1, 6)],
        "E": [i for i in range(1, 6)],
    } for movie_num in movies.keys()
    }

# Display the menu
def main():
    while True:     # Menu of the program and prompt the user to enter a choice
        print("\nWelcome to Cinema Booking System!")
        print("1. View Showings")
        print("2. Reserve Ticket")
        print("3. Exit Menu")
        choice = input("Enter choice (1-3): ")
        if choice == "1":       # Calls the function depending on the user's choice
            view_movies()
        elif choice == "2":
            reserve_ticket()
        elif choice == "3":
            print("\nThank you for using Cinema Booking System!")
            break
        else:
            print("Invalid choice, please try again.")

# Display avaliable seats for a particular movie
def print_seats(movie):
    print(f"\nAvailable seats for {movies[movie]['title']}:")
    table_data = [] 
    for row, nums in seats[movie].items():      # Loop through each seat number and store them in the table_data
        table_data.append([row] + [f"{num}" if num in nums else "   " for num in range(1, 6)])
    headers = [""] + [f"{i}" for i in range(1, 6)]
    print(tabulate(table_data, headers, tablefmt="fancy_grid")) 

# Display available movies
def view_movies():
    print("\nAvailable movies:")
    headers = ["  ", "Title", "Genre", "Price"]
    rows = []
    for key in movies.keys():
        title = movies[key]["title"]
        genre = movies[key]["Genre"]
        price = movies[key]["price"]        
        rows.append([key, title, genre, price])
    print(tabulate(rows, headers=headers, tablefmt="fancy_grid"))

# Function to reserve a ticket
def reserve_ticket():
    print(" ")      # Display list of movies and prompt user to select one
    print("\nPlease select a movie to book tickets for:")
    view_movies()
    movie = input("Enter movie number: ")
    while movie not in movies:
        print("Invalid movie number, please try again.")
        movie = input("Enter movie number: ")
    print(f"Selected movie: {movies[movie]['title']}")
    
    print_seats(movie)      # Ask user to select a seat and confirm selection
    seat = input("Enter row (A-E): ").upper()
    while seat not in seats[movie]:
        print("Invalid row, please try again.")
        seat = input("Enter row (A-E): ").upper()
    num = int(input("Enter seat number (1-5): "))
    while num not in seats[movie][seat]:
        print("Seat is not available, please choose another seat.")
        num = int(input("Enter seat number (1-5): "))
    print(" ")
    print(f"Selected seat: {seat}{num}")
    
    print(" ")      # Confirm the booking and display booking details
    confirm = input("Confirm booking? (Y/N): ").upper()
    while confirm not in ["Y", "N"]:
        print("Invalid option, please try again.")
        confirm = input("Confirm booking? (Y/N): ").upper()
    if confirm == "Y":
        seats[movie][seat][num-1] = " "
        generate_receipt(movie, seat, num)
        print(" ")
        print("Booking confirmed!")
    else:
        print("Booking canceled.")

# Generate receipt with random transaction number
def generate_receipt(movie, seat, num):
    transaction_number = random.randint(100000000, 999999999)
    print("\nYou made a reservation for: ")
    table = [["Transaction No:", transaction_number],
             ["Movie:", movies[movie]['title']],
             ["Seat:", seat+str(num)],
             ["Price:", f"₱{movies[movie]['price']}"]]
    print(tabulate(table, tablefmt="fancy_grid"))
    save_bookings(movie, seat, num, transaction_number)

# Save bookings in an Excel file
def save_bookings(movie, seat, num, transaction_number):
    try:
        wb = load_workbook("bookings.xlsx")
    except FileNotFoundError:
        wb = Workbook()
    sheet = wb.active
    if sheet.max_row == 0:
        sheet["A1"] = "Movie"
        sheet["B1"] = "Seat"
        sheet["C1"] = "Price"
        sheet["D1"] = "Transaction No."
    row = sheet.max_row + 1
    sheet.cell(row=row, column=1, value=movies[movie]["title"])
    sheet.cell(row=row, column=2, value=f"{seat}{num}")
    sheet.cell(row=row, column=3, value=f"₱{movies[movie]['price']:.2f}")
    sheet.cell(row=row, column=4, value=transaction_number)
    wb.save("bookings.xlsx")

main()

File: JAVA/CinemaBookingSystem.java
import java.util.*;


public class CinemaBookingSystem {
    static Map<String, Map<String, List<Integer>>> seats = new HashMap<>();
    static Map<String, Map<String, Object>> movies = new HashMap<>();
    static {
        movies.put("1", new HashMap<String, Object>() {{
            put("title", "Ant-Man And The Wasp: Quantumania");
            put("Genre", "Action/Adventure");
            put("price", 680);
        }});
        movies.put("2", new HashMap<String, Object>() {{
            put("title", "Avatar: The Way Of Water");
            put("Genre", "Action/Adventure");
            put("price", 680);
        }});
        movies.put("3", new HashMap<String, Object>() {{
            put("title", "Sound Of Silence");
            put("Genre", "Action/Adventure");
            put("price", 540);
        }});
        movies.put("4", new HashMap<String, Object>() {{
            put("title", "The First Slam Dunk");
            put("Genre", "Anime");
            put("price", 480);
        }});
        movies.put("5", new HashMap<String, Object>() {{
            put("title", "Oras De Peligro");
            put("Genre", "Drama");
            put("price", 380);
        }});

        for (String movie_num : movies.keySet()) {
            Map<String, List<Integer>> rows = new HashMap<>();
            rows.put("A", Arrays.asList(1, 2, 3, 4, 5));
            rows.put("B", Arrays.asList(1, 2, 3, 4, 5));
            rows.put("C", Arrays.asList(1, 2, 3, 4, 5));
            rows.put("D", Arrays.asList(1, 2, 3, 4, 5));
            rows.put("E", Arrays.asList(1, 2, 3, 4, 5));
            seats.put(movie_num, rows);
        }
    }
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        while (true) {
            System.out.println("\nWelcome to Cinema Booking System!");
            System.out.println("1. View Showings");
            System.out.println("2. Reserve Ticket");
            System.out.println("3. Exit Menu");

            System.out.print("Enter choice (1-3): ");
            String choice = sc.nextLine();
            switch (choice) {
                case "1":
                    viewMovies();
                    break;
                case "2":
                    reserveTicket(sc);
                    sc.nextLine();
                    break;
                case "3":
                    System.out.println("\nThank you for using Cinema Booking System!");
                    return;
                default:
                    System.out.println("Invalid choice. Please try again.");
            }
        }
    }
    public static void viewMovies() {
        System.out.printf(" -----------------------------------------------------------------------%n");
        System.out.printf("|                       N O W   S H O W I N G                           |%n");
        System.out.printf(" -----------------------------------------------------------------------%n");
        for (String key : movies.keySet()) {
            String title = (String) movies.get(key).get("title");
            String genre = (String) movies.get(key).get("Genre");
            int price = (int) movies.get(key).get("price");
            System.out.printf("| %-1s | %-35s | %-17s | %-8s|%n", key, title, genre, price);
        }
        
        System.out.printf(" -----------------------------------------------------------------------%n");

   
}

    public static void reserveTicket(Scanner sc) {
        System.out.println("\nPlease select a movie to book tickets for:");
        viewMovies();
        System.out.print("Enter movie number (1-5): ");
        String movie_num = sc.nextLine();
        if (!movies.containsKey(movie_num)) {
            System.out.println("Invalid movie number. Please try again.");
            return;
        }
        String movie_title = (String) movies.get(movie_num).get("title");
        int price = (int) movies.get(movie_num).get("price");
    
        System.out.print("\nEnter number of tickets to book: ");
        int num_tickets = sc.nextInt();
        if (num_tickets <= 0) {
            System.out.println("Invalid number of tickets. Please try again.");
            return;
        }
    
        Map<String, List<Integer>> rows = seats.get(movie_num);
        System.out.println("\nAvailable seats for the movie:");
        System.out.println("----+-----+-----+-----+-----+-----+");
        System.out.println("    |  1  |  2  |  3  |  4  |  5  |");
        System.out.println("====+=====+=====+=====+=====+=====+");
        for (String row : rows.keySet()) {
            System.out.printf("%s   |", row);
            for (int i = 1; i <= 5; i++) {
                System.out.printf("%2s   |", rows.get(row).contains(i) ? i : "  ");
            }
            System.out.println();
            System.out.println("----+-----+-----+-----+-----+-----+");
        }
        List<String> selected_seats = new ArrayList<>();
        for (int i = 1; i <= num_tickets; i++) {
            System.out.printf("\nEnter seat number for ticket %d (e.g. A1): ", i);
            String seat = sc.next().toUpperCase();
            String row = seat.substring(0, 1);
            int col = Integer.parseInt(seat.substring(1));
            if (!rows.containsKey(row) || !rows.get(row).contains(col)) {
                System.out.println("Invalid seat number. Please try again.");
                continue;
            }
            selected_seats.add(seat);
        }

        for (String seat : selected_seats) {
            String row = seat.substring(0, 1);
            int col = Integer.parseInt(seat.substring(1));
            List<Integer> cols = new ArrayList<>(rows.get(row));
            cols.remove(Integer.valueOf(col));
            rows.put(row, cols);
        }
        int total_price = price * num_tickets;

        System.out.printf(" ------------------------------------------------%n");
        System.out.printf("|            You made a reservation for          |%n");
        System.out.printf(" ------------------------------------------------%n");
        System.out.printf("| %-8s | %-35s |%n", "Movie", movie_title);
        System.out.printf("| %-8s | %-35s |%n", "Seat", selected_seats);
        System.out.printf("| %-8s | %-35s |%n", "Price", total_price);
        System.out.printf(" -------------------------------------------------%n");

        System.out.print("\nDo you want to confirm the booking? (Y/N): ");
        String confirm = sc.next().toUpperCase();

        if (confirm.equals("Y")) {
            System.out.println(" ");
            System.out.println("Booking confirmed!");
            System.out.println("Thank you for using Cinema Booking System!");
        } 
        
    }
    
}